Sources/FoundationInternationalization/Formatting/Number/BinaryInteger+FormatStyle.swift

```diff
@@ -34,6 +40,199 @@ extension BinaryInteger {
+        //
+        // So we replace some of those expensive O(log2(bitWidth)) divides with simpler O(1) divides, by first dividing by the largest multiple of ten such that the remainder fits in a single machine word (UInt), and then using regular integer division CPU instructions to further divide that simple machine-word-sized remainder down into individual digits.
+
+        let (decimalDigitsPerWord, wordMagnitude) = Self.decimalDigitsAndMagnitudePerWord()
+        let negative = 0 > self
+        let maximumDigits = (Self.maximumDecimalDigitsForUnsigned(bitWidth: self.bitWidth - (negative ? 1 : 0)) // -1 for negative values because their bit width includes the sign bit, which we don't care about.
+                             + (negative ? 1 : 0)) // Include room for "-" prefix if necessary.
+        var actualDigits: Int = Int.min // Actually initialised inside the closure below, but the compiler mistakenly demands a default value anyway.
+
+        return ContiguousArray<UInt8>(unsafeUninitializedCapacity: maximumDigits) { buffer, initialisedCount in
+            var tmp = self
+            var wordInsertionPoint = buffer.endIndex - 1
+
+            while .zero != tmp {
+                let (quotient, remainder) = tmp.quotientAndRemainder(dividingBy: wordMagnitude)
+                precondition(.zero == remainder || (negative == (0 > remainder)), "Starting value \(tmp) is \(negative ? "negative" : "positive (or zero)") yet the remainder of division by \(wordMagnitude) is not: \(remainder).  quotientAndRemainder(dividingBy:) is not implemented correctly for \(type(of: self)) (it might be using T-division instead of F-division).") // It's an entirely understandable error for an implementor to use T-division for their integer quotient and remainder, but they're supposed to use F-division.  i.e. the quotient is supposed to be rounded towards zero rather than down (and that effects the modulus correspondingly, since either way the results must satisfy r = d ⨉ (r idiv i) + (r mod i)).  F-division is convenient because its remainder is neatly the value of interest to this algorithm, rather than being offset by the divisor if r is negative.  While it would be technically possible to assume T-division if the remainder's sign doesn't match, the incorrect implementation of quotientAndRemainder(dividingBy:) will probably still break other algorithms and so we shouldn't encourage it.
+
+                // By definition the remainder has to be a single word (since the divisor, `wordMagnitude`, fits in a single word), so we can avoid working on a BinaryInteger generically and just use the first word directly, which is concretely UInt.
+                assert(remainder.bitWidth - (negative ? 1 : 0) <= Words.Element.bitWidth,
```

Comment:
The bit-width of fixed-width integers is not normalized, so `remainder.bitWidth` may be fixed to a value larger than `UInt.bitWidth` (see `Int256`).

Reply:
Thanks for catching that.

This would presumably also make Int64 & UInt64 fail in 32-bit mode.

It's very unfortunate that it's documented to behave that way; it seems to render that property useless.  The full width of fixed-width integers is already provided by the type-level property `bitWidth`.

The claimed bit width is relied upon for more than just that assert; I'll have to think a bit about how to work around this problem…

Reply:
…which, it turns out, is not easy.  I'm still trying to find a way; no luck today, I'll have to pick this up again some other day.

Mathematically it's trivial, but because of the structure of the `BinaryInteger` and associated types (in particular that `Magnitude` is merely `Numeric`, which is virtually useless) it is hard to actually implement the necessary functionality.

Though in looking at this I realised the size estimation, based on bit width, was technically wrong anyway.  Even when `bitWidth` returns what I was expecting.  I don't _think_ it could ever actually cause an off-by-one bug because I think it would require a power of ten and a power of two to be equal (which is impossible), but still, it'd be nice to eliminate that possibility as a matter of principle.  I'm still looking into that, too.

Reply:
Okay, life found a way.  The end result isn't that complicated, actually, and might even be more efficient than before since we can take advantage of the fact that we only care about the _magnitude_, really, as that's what we're essentially converting to decimal digits.  So a lot of the complexities of two's complement representation can be side-stepped.

A positive coincidental side-effect of this is that the algorithm no longer relies on the `bitWidth` instance property at all (though still the type-level version for UInt, which is fine).  That lack of reliance is is good because it turns out some 3rd party implementations of `BinaryInteger` don't implement it correctly anyway (e.g. https://github.com/attaswift/BigInt/issues/115).

Reply:
Oh, and also, [attaswift/BigInt](https://github.com/attaswift/BigInt) conforms both its types (`BigInt` and `BigUInt`) to `FixedWidthInteger` even though they're not fixed width.  Perhaps just to make it more convenient to use (I actually [ab]use this in the unit tests here, to use the `init` method that takes a string with configurable radix)?  So my initial attempts to special-case `bitWidth` interpretation for non-`FixedWidthInteger` `BinaryInteger`s was doomed to fail anyway (or, at least, to be irrelevant).